month2_1.hs

```diff
@@ -0,0 +1,106 @@
+import Prelude hiding ((!!),init,reverse,(++),cycle,take,elem)
+--1
+(!!)::[a] -> Int -> a
+(!!) xs n | n < 0 = error "Element should be positive!"
+(!!) [] _         = error "Element out of list!"
+(!!) xs 0         = head xs
+(!!) xs n         = (!!) (tail xs) (n-1)
+--['a'..'z'] !! 5
+
+--2
+init :: [a] -> [a]
+init []           = error "Empty list!"
+init [x]          = []
+init (x:xs)       = x:(init xs)
+--init [0..6]
+
+--3
+(++) :: [a] -> [a] -> [a]
+(++) [] ys        = ys
+(++) (x:xs) ys    = x : (++) xs ys
+-- [1..5] ++ [6..9]
+
+--4
+cycle :: [a] -> [a]
+cycle xs = (++) xs (cycle xs)
+-- take 25  cycle [1..5]
+
+--5
+take:: Int -> [a] -> [a]
+take n _ | n < 0  = error "number should be positive"
+take n []         = []
+take n xs         = take' n xs where
+    take' n []      = error "not enough elements in the list"
+    take' 0 xs      = []
+    take' n (x:xs)  = x : (take' (n - 1) xs)
+--take 50 [1..]
+
+--6
+inits:: [a] -> [[a]]
+inits []        = error "Empty list"
+inits [x]       = [[], [x]]
+inits xs        = (++) (inits $ init xs) [xs]
+--inits ['a'.. 'h']
+
+tails:: [a] -> [[a]]
+tails []        = error "Empty list"
+tails [x]       = [[x], []]
+tails xs        = (++) [xs] (tails $ tail xs)
+--tails [1..7]
+
+--7
+elem :: Eq a => a -> [a] -> Bool
+elem _ []                     = False
+elem n (x:xs) | n == x        = True
+              | otherwise     = elem n xs
+--elem 'd' ['1'..'9']
+--elem 'd' ['a'..'f']
+
+--8
+nub:: Eq a => [a] -> [a]
+nub []                           = []
+nub (x:xs) | elem x xs == True   = nub xs
+           | elem x xs == False  = (++) [x] (nub xs)
```

Comment:
`elem x xs == True`  `<=>` `elem x xs`, во втором  случае можно писать otherwise. Здесь значения особого нет, но если бы elem был дорогой по вычислению, можно было бы здорово ухудшить себе жизнь